### PR TITLE
WIP: Support "stay open" mode for Lambda invocations

### LINKED ISSUE #4123: Support "stay open" mode for Lambda invocations
# Type of request: This is a ...

- [ ] bug report
- [x] feature request

# Detailed description

I'm currently using Localstack for local integration + performance tests, to validate the quality of my local Lambda, before it gets shipped to AWS.

Something I've noticed is that, when running Lambdas with languages that perform less well on cold starts (such as Java), the performance of these Lambdas are not ideal.

I'd originally seen this when using the `local` executor, so thought that moving to the `docker` or `docker-reuse` commands would speed this up, but it appears that the container is not actually re-used across invocations, and is deleted each time.

(This may be exacerbated by the fact that we only allow a single Lambda invocation - this may be an additional feature request to add?)

This could be avoided by using the [configuration available in `lambci/lambda`](https://github.com/lambci/docker-lambda#running-in-stay-open-api-mode) to "stay open" allowing the Lambda to stay running, and only be closed when the Localstack container is being closed.

## Expected behavior

Repeat invocations of a Lambda does not need a cold start, as the Lambda will be pre-warmed and re-used.

## Actual behavior

A cold start occurs on each Lambda invocation.

# Steps to reproduce

Localstack version: `0.12.9`

## Command used to start LocalStack

`docker-compose up` via https://gitlab.com/jamietanna/localstack-docker-reuse-not-staying-open/-/blob/main/docker-compose.yml

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

https://gitlab.com/jamietanna/localstack-docker-reuse-not-staying-open

Notice that cold starts occur each time (in the constructor of `Handler`)

### PR DESCRIPTION
**Pull request by chaz-doyle-cko**
Fix to `LAMBDA_EXECUTOR=docker-reuse` by using the `DOCKER_LAMBDA_STAY_OPEN` in lambci/lambda containers to allow faster lambda invocations.

This as been tested against both `nodejs` and `.NET` lambdas (`.NET` using custom runtime `provided.al2`)

Please feel free test further @jamietanna, @wesselvdv

**NOTE:** I had to fork brand new as the current changes in `master` are extensive enough to warrant a clean slate.

[Replaces PR-4185](https://github.com/localstack/localstack/pull/4185)
[Closes #4123](https://github.com/localstack/localstack/pull/4185#issue-927174042)

### CODE REVIEW

**Comment by jamietanna** [comment]
I've validated this with the blank-java project, and can see that cold starts are not present after the intiial call :raised_hands: Awesome stuff, @chaz-doyle-cko!

**Comment by barnc** [comment]
You legend! This is huge for our team - massive thanks. 👍

**Review by whummer — COMMENTED** [review]
Great, thanks for digging into this @chaz-doyle-cko  @jamietanna @barnc  ! 👍 

Looks like the tests haven't been triggered on this PR - usually this can be fixed if you log in to CircleCI with your Github credentials to re-initialize the repo permissions - can you please give that a try? (hope that helps..)

I've left a few initial comments - perhaps it would be easier if I quickly raise a PR with these changes  against your fork, or simply create a new/updated PR in this repo (we'll make sure that your authorship is retained..). Thanks!

File: localstack/services/awslambda/lambda_executors.py [hunk 1]
```diff
@@ -783,14 +789,47 @@ def execute_in_container(
                 container_info.name, f"{lambda_cwd}/.", DOCKER_TASK_FOLDER
             )
```
[the remark this answers is not in the document]
Comment (chaz-doyle-cko) [reply localstack/services/awslambda/lambda_executors.py:790]: Completely agree... Have updated

**Comment by chaz-doyle-cko** [comment]
Superseded by [4914](https://github.com/localstack/localstack/pull/4914) (building on the work in this PR)

**Pull request closed without merging**